File: Finished And Submitted/Binary Tree Level Order Traversal.go
package todos

import "leetcode/ds"

// https://leetcode.com/problems/binary-tree-level-order-traversal/

func levelOrder(root *ds.TreeNode) [][]int {
	return levelOrderHelper(root, 0, [][]int{})
}

func levelOrderHelper(node *ds.TreeNode, depth int, output [][]int) [][]int {
	if node == nil {
		return output
	}

	if len(output) >= depth+1 {
		output = append(output, []int{})
	}

	output[depth] = append(output[depth], node.Val)
	output = levelOrderHelper(node.Left, depth+1, output)
	return levelOrderHelper(node.Right, depth+1, output)
}

// func levelOrder(root *ds.TreeNode) [][]int {
// 	if root == nil {
// 		return [][]int{}
// 	}

// 	left := levelOrder(root.Left)
// 	right := levelOrder(root.Right)

// 	result := [][]int{[]int{root.Val}}
// 	for i := 0; i < maxInt(len(left), len(right)); i++ {
// 		if i >= len(left) {
// 			result = append(result, right[i])
// 		} else if i >= len(right) {
// 			result = append(result, left[i])
// 		} else {
// 			result = append(result, append(left[i], right[i]...))
// 		}
// 	}

// 	return result
// }

// func maxInt(a, b int) int {
// 	if a > b {
// 		return a
// 	}
// 	return b
// }

File: src/python/finished/binary_tree_level_order_traversal.py
# https://leetcode.com/problems/binary-tree-level-order-traversal/

# Definition for a binary tree node.

from collections import deque
from typing import List, Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        dq = deque()
        dq.append((0, root))
        output = []

        while len(dq) > 0:
            (l, n) = dq.popleft()

            if n == None:
                continue
            
            while len(output) <= l:
                output.append([])

            dq.append((l+1, n.left))
            dq.append((l+1, n.right))
            output[l].append(n.val)
        
        return output
